### PR TITLE
Update Ember types for Ember v4

### PR DESCRIPTION
**Pull request by chriskrycho**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <[<url here>](https://deprecations.emberjs.com/v3.x/)>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

## Notes

This PR represents an update to *all* the types under the `ember` and `@ember/*` packages to bring them in line with Ember v4, in particular by removing all deprecations from the 3.x era, and by supplying the `v3` types in their own folder per the usual DefinitelyTyped flow.

I also made a number of additional changes in support of this:

- I updated the minimum required TypeScript version (MSTSV!) to 4.4, in line with the policy we expect to adopt for TypeScript support with Ember itself in the months ahead.

- Since this is already a breaking change, I made a number of other updates which we have known needed doing for quite some time:
    - updated all the `tsconfig.json` files to require `strict: true` and fixed all code broken by that;
    - replaced `any` with `unknown` as appropriate throughout.

- I have fixed several bugs I found along the way, including a number of type definitions which appear to have been wrong *the entire history of these types*. 😱 

- I have removed previous maintainers who have been inactive for a while, just to stop bothering them with notifications! Many thanks to all those previous maintainers, but especially to @mike-north and @dwickern, whose invaluable contributions are a significant part of the reason these types exist and work well today! (Previous maintainers, if you *want* to stay on the list, I’m happy to add you back to it.)

This nicely lays the groundwork for porting these types into Ember itself.

**Note:** this does *not* update the types for Ember Data beyond bumping the version number to 4.0 and MSTSV to match Ember itself. (I do not have the expertise to do that, but will happily support others in doing that work! The goal here was to *unblock* it for others, by getting the gnarly foundational bits out of the way.)

### CODE REVIEW

File: types/ember__component/index.d.ts [hunk 1]
```diff
@@ -93,3 +94,17 @@ export default class Component extends CoreView.extend(
...
+/**
+ * Associate a class with a component manager (an object that is responsible for
+ * coordinating the lifecycle events that occurs when invoking, rendering and
+ * re-rendering a component).
+ *
+ * @param managerFactory a function to create the owner for an object
+ * @param object the object to associate with the componetn manager
+ * @return the same object passed in, now associated with the manager
+ */
+export function setComponentManager<T>(
```
Comment (dfreeman) [inline types/ember__component/index.d.ts:106]: Do we need to declare the `capabilities` export as well? As far as I know that's the only documented(?) way to produce a `ComponentCapabilities` object, and without it I'm not sure how authors could get their hands on a value that would be legal to use with `setComponentManager` as declared here

File: types/ember/index.d.ts [hunk 2]
```diff
@@ -153,15 +153,6 @@ export namespace Ember {
     class Resolver extends EmberResolver {}
     class EventDispatcher extends EmberEventDispatcher {}
     class Registry extends EmberRegistry {}
-    interface FunctionPrototypeExtensions {
```
Comment (dfreeman) [inline types/ember/index.d.ts:164]: 🎉

File: types/ember/index.d.ts [hunk 3]
```diff
@@ -151,17 +151,6 @@ export namespace Ember {
     class Registry extends EmberRegistry {}
     interface ArrayPrototypeExtensions<T> extends EmberArrayProtoExtensions<T> {}
 
-    interface StringPrototypeExtensions {
```
Comment (dfreeman) [inline types/ember/index.d.ts:187]: 🎉

File: types/ember/index.d.ts [hunk 4]
```diff
@@ -379,8 +379,20 @@ export namespace Ember {
...
-    const run: typeof EmberRunloopNs.run;
+    const computed: typeof EmberObjectNs.computed & typeof EmberObjectComputedNs;
+
+    // Shenanigans to make `run` both callable and a namespace safely, while not
+    // making the `run.bind` call resolve to `Function.prototype.bind`. (Yes,
+    // insert :upside-down-smiley: here.)
+    // 1. Get the type side of the namespace.
+    type EmberRunloop = typeof EmberRunloopNs;
+    // 2. Use it to get an interface representing the callable side of `run`.
+    type RunFn = Pick<EmberRunloop, 'run'>['run'];
```
Comment (dfreeman) [inline types/ember/index.d.ts:311]: I think you can skip the `Pick` here?

```ts
type RunFn = EmberRunloop['run'];
```

File: types/ember-qunit/ember-qunit-tests.ts [hunk 5]
```diff
@@ -7,159 +6,132 @@ import {
...
     todo,
-    moduleFor,
-    moduleForModel,
-    moduleForComponent,
     setResolver,
     setupRenderingTest,
     setupTest,
 } from 'ember-qunit';
-import { TestContext } from 'ember-test-helpers';
+import { render, TestContext } from '@ember/test-helpers';
```
Comment (dfreeman) [inline types/ember-qunit/ember-qunit-tests.ts:13]: The `TestContext` move is going to be a fun one to deal with in consuming codebases 🥲

**Comment by dfreeman** [comment]
I can't claim to have looked at the big moving-things-around commits in great detail, but overall this all looks reasonable to me, and it includes a handful of fixes/cleanups I've been wanting for a _long_ time. Thank you for your Herculean effort here, @chriskrycho!

File: types/ember-qunit/ember-qunit-tests.ts (hunk 5 above)
Comment (chriskrycho) [reply types/ember-qunit/ember-qunit-tests.ts:13]: Alas, yes. I thought about re-exporting, but at some point people are going to have to bite the bullet, and no one should actually be using `ember-test-helpers` at runtime anymore, so… 🤷🏼‍♂️

File: types/ember__component/index.d.ts (hunk 1 above)
Comment (chriskrycho) [reply types/ember__component/index.d.ts:106]: Mmmmm, probably. I’m poking at it and I don’t actually know how to expose some of this (it’s not really exposed in a clean way in Ember’s or Glimmer’s internal types, such that it’s not clear what’s intended to be public and what isn’t). Will check with folks who’d know.

File: types/ember/index.d.ts (hunk 4 above)
Comment (chriskrycho) [reply types/ember/index.d.ts]: Yeah, there was some reason we(/I?) did this that had to do with how type resolution was working at some point (the `Pick` *was* actually making a difference?), but I reran all the relevant tests and it's fine now.

**Comment by chriskrycho** [comment]
Note for reviewer (and bot!) I’ve updated the `tsconfig.json` values for the *new* packages to move toward expected, but have intentionally *not* touched the values for the newly-introduced previous-version packages (for what I trust are obvious reasons).

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ember-feature-flags ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-feature-flags@latest/))
was missing the following properties:
1. name
1. name
## ember-mocha ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-mocha@latest/))
was missing the following properties:
1. name
1. init
1. postBuild
1. checkPackages
1. included

<details>
<summary>as well as these 17 other properties...</summary>
<p>targetOptions, contentFor, treeForVendor, treeForAddonTestSupport, overrideTestCommandFilter, setTestGenerator, name, init, postBuild, checkPackages, included, targetOptions, contentFor, treeForVendor, treeForAddonTestSupport, overrideTestCommandFilter, setTestGenerator</p>
</details>

## ember-qunit ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-qunit@latest/))
was missing the following properties:
1. name
1. init
1. included
1. targetOptions
1. treeForVendor

<details>
<summary>as well as these 30 other properties...</summary>
<p>treeForAddonTestSupport, setTestGenerator, name, init, included, targetOptions, treeForVendor, treeForAddonTestSupport, setTestGenerator, name, init, included, targetOptions, treeForVendor, treeForAddonTestSupport, setTestGenerator, name, init, included, targetOptions, treeForVendor, treeForAddonTestSupport, setTestGenerator, name, init, included, targetOptions, treeForVendor, treeForAddonTestSupport, setTestGenerator</p>
</details>

## ember-resolver ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-resolver@latest/))
was missing the following properties:
1. name
1. emberResolverFeatureFlags
1. init
1. treeForAddon
1. included

<details>
<summary>as well as these 7 other properties...</summary>
<p>monkeyPatchVendorFiles, name, emberResolverFeatureFlags, init, treeForAddon, included, monkeyPatchVendorFiles</p>
</details>

## ember__test-helpers ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember__test-helpers@latest/))
was missing the following properties:
1. name
1. treeForAddonTestSupport
1. name
1. treeForAddonTestSupport
1. name

<details>
<summary>as well as these 5 other properties...</summary>
<p>treeForAddonTestSupport, name, treeForAddonTestSupport, name, treeForAddonTestSupport</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against c3d9c617461eb133da224327dfb41b5b74015a5f
</p>

**Comment by chriskrycho** [comment]
Lulz. The bot *lies* about what the packages should have. Okay. Pushing a fix for that now.

**Review by dfreeman — APPROVED** [review]
:shipit:

**Pull request merged**